### PR TITLE
Expose kernel  API

### PR DESCRIPTION
**Pull request by DonJayamanne**
For #5532

* [x] Allow anyone to use this in debug mode
* [x] Add telemetry for usage

### CODE REVIEW

File: src/client/api/extension.d.ts [hunk 1]
```diff
@@ -0,0 +1,256 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
```
Comment (DonJayamanne) [inline src/client/api/extension.d.ts:1]: Later we can publish this as npm or github package, else users have to hand craft this.
For now lets get users to manually copy this over.

File: src/client/api/extension.d.ts [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+     */
+    interpreterPath?: string;
+    readonly interrupt_mode?: 'message' | 'signal';
+}
+/**
+ * Connection metadata for Kernels started using kernelspec (JSON).
+ * This could be a raw kernel (spec might have path to executable for .NET or the like).
+ * If the executable is not defined in kernelspec json, & it is a Python kernel, then we'll use the provided python interpreter.
+ */
+export type LocalKernelSpecConnectionMetadata = Readonly<{
```
Comment (rchiodo) [inline src/client/api/extension.d.ts:102]: I'm confused, this is exported here and in types.ts? Which one do we want to expose to people?

**Comment by rchiodo** [comment]
Looks good, just had some questions about where the KernelConnectionMetadata types are defined (seems like they're in more than one spot)

File: src/client/api/extension.d.ts (hunk 2 above)
Comment (DonJayamanne) [reply src/client/api/extension.d.ts:102]: Users of this API will need the *.d.ts files (or the types).
This file contains everything in one place, this way they can just get it from here.
Ideally we should push this as an npm package or the like, but we can do that later

File: src/client/api/apiAccessService.ts [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+    extensionId: string;
+    allowed: 'yes' | 'no';
+}[];
+
+const API_ACCESS_GLOBAL_KEY = 'JUPYTER_API_ACCESS_INFORMATION';
+// Some publishers like our own `ms-tolsai` will always be allowed to access the API.
+const TrustedExtensionPublishers = new Set([JVSC_EXTENSION_ID.split('.')[0]]);
+// We dont want to expose this API to everyone, else we'll keep track of who has access to this.
+// However, the user will still be prompted to grant these extensions access to the kernels..
+const PublishersAllowedWithPrompts = new Set([JVSC_EXTENSION_ID.split('.')[0]]);
```
[the remark this answers is not in the document]
Comment (DonJayamanne) [reply src/client/api/apiAccessService.ts:22]: lets discuss in standup, when Ian is also back

File: src/client/api/extension.d.ts (hunk 2 above)
Comment (rchiodo) [reply src/client/api/extension.d.ts:102]: But doesn't that mean the other location isn't necessary? Can't we just use this definition ourselves?

File: src/client/api/apiAccessService.ts [hunk 4]
```diff
@@ -29,11 +29,20 @@ export class ApiAccessService {
...
+        @inject(IApplicationEnvironment) private appEnv: IApplicationEnvironment,
+        @inject(IExtensionContext) private context: IExtensionContext
     ) {}
     public async getAccessInformation(): Promise<{ extensionId: string; accessAllowed: boolean }> {
         const info = await this.extensions.determineExtensionFromCallStack();
         const publisherId = info.extensionId.split('.')[0];
+        if (this.context.extensionMode === ExtensionMode.Test) {
+            if (!TrustedExtensionPublishers.has(publisherId) || PublishersAllowedWithPrompts.has(publisherId)) {
+                void this.appShell.showInformationMessage(
+                    'Thanks for trying the Jupyter API. Please file an issue on our repo to use this API in production. This would prevent us from breaking your extension when updating the API (as it is still a work in progress).'
```
Comment (rchiodo) [inline src/client/api/apiAccessService.ts:41]: This needs to be localized.

File: src/client/api/extension.d.ts (hunk 2 above)
Comment (DonJayamanne) [reply src/client/api/extension.d.ts:102]: Not really, I don't want to expose all of the properties, I don't want to expose everythign in Python interpreter or some other properties, else we end up in advertently having to support all of the stuff we expose.
This way we have full control over what's exposed.

Comment (DonJayamanne) [reply src/client/api/extension.d.ts:102]: E.g. I don't want to expose everything in the Python interpretre API

File: src/client/api/apiAccessService.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (DonJayamanne) [reply src/client/api/apiAccessService.ts]: My concern is anyone can edit the package.json, i prefer the stuff hardcoded
Though I do like the package.json approach as well, feels a lot cleaner.
But something about that keeps bothering me, feels like we're just dumping stuff into that file, then again, there's no reason we can't put anything we want to...
As you can tell i have no preference for either one

**Review by rchiodo — APPROVED** [review]

**Pull request merged**